GCC <=12 is incompatible - CI is biased

Hi,

I tried to compile a container image with the module from `valkey/valkey:8` (not yet on alpine).

A lot of errors happen.

I fixed this by downgrading GCC and G++ to v11.

Here is the Dockerfile that works:

```dockerfile
FROM docker.io/valkey/valkey:8 as build

ARG SERVER_VERSION=8.0.0
RUN set -ex; \
  apt-get update; \
  apt-get install -y build-essential cmake git gcc-11 g++-11 pkg-config; \
  update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-11 1; \
  update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-11 1; \
  mkdir -p /opt;

WORKDIR /opt
RUN set -ex; \
  git clone https://github.com/valkey-io/valkeyJSON.git; \
  mkdir -p /opt/valkeyJSON/build;

WORKDIR /opt/valkeyJSON/build
RUN set -ex; \
  cmake .. -DVALKEY_VERSION=${SERVER_VERSION}; \
  make

FROM docker.io/valkey/valkey:8
RUN mkdir -p /usr/lib/valkey
COPY --from=build /opt/valkeyJSON/build/src/libjson.so /usr/lib/valkey/libjson.so
CMD ["valkey-server", "--loadmodule", "/usr/lib/valkey/libjson.so"]
```

On Github Actions, Ubuntu version is not the latest, and not the one used by the main OCI Image. So the tests are OK on GitHub actions, but not with the official image.

Had a similar issue on debian bookworm

root@dev303-valkey:/opt/valkeyJSON/bld# cmake .. -DVALKEY_VERSION=8.0.2
Building valkeyJSON for x86_64
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
Valkey version: 8.0.2
CFLAGS: -g -O3 -fno-omit-frame-pointer -Wall -Werror -Wextra
CMAKE_C_FLAGS:  -g -O3 -fno-omit-frame-pointer -Wall -Werror -Wextra -fPIC -march=nehalem
CMAKE_CXX_FLAGS:  -g -O3 -fno-omit-frame-pointer -Wall -Werror -Wextra -fPIC -march=nehalem
CMake Deprecation Warning at bld/_deps/rapidjson-src/CMakeLists.txt:1 (CMAKE_MINIMUM_REQUIRED):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


src/CMakeLists.txt: Build valkeyJSON
tst/CMakeLists.txt
-- Found Python: /usr/bin/python3 (found version "3.11.2") found components: Interpreter 
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
tst/unit/CMakeLists.txt: Define unit tests
-- Configuring done
-- Generating done
-- Build files have been written to: /opt/valkeyJSON/bld


root@dev303-valkey:/opt/valkeyJSON/bld# ls -lrt
total 64
drwxr-xr-x 10 root root  4096 Feb 13 00:56 _deps
-rw-r--r--  1 root root 23357 Feb 13 00:56 CMakeCache.txt
-rw-r--r--  1 root root  2092 Feb 13 00:56 cmake_install.cmake
-rw-r--r--  1 root root 11606 Feb 13 00:56 Makefile
drwxr-xr-x  3 root root  4096 Feb 13 00:56 src
drwxr-xr-x  4 root root  4096 Feb 13 00:56 tst
drwxr-xr-x  2 root root  4096 Feb 13 00:56 lib
drwxr-xr-x  2 root root  4096 Feb 13 00:56 bin
drwxr-xr-x  8 root root  4096 Feb 13 00:56 CMakeFiles


root@dev303-valkey:/opt/valkeyJSON/bld# make
[  2%] Creating directories for 'valkey'
[  4%] Performing download step (git clone) for 'valkey'
Cloning into 'valkey'...
HEAD is now at d6a2f561e Update release notes with final security information
[  6%] Copying header files to include/ directory
[  8%] Performing update step for 'valkey'
[ 10%] No patch step for 'valkey'
[ 13%] No configure step for 'valkey'
[ 15%] Performing build step for 'valkey'
/bin/sh: 1: pkg-config: not found
/bin/sh: 1: pkg-config: not found
/bin/sh: 1: pkg-config: not found
/bin/sh: 1: pkg-config: not found
/bin/sh: 1: pkg-config: not found
/bin/sh: 1: pkg-config: not found
    CC Makefile.dep
/bin/sh: 1: pkg-config: not found
/bin/sh: 1: pkg-config: not found
/bin/sh: 1: pkg-config: not found
    GEN commands.def
MAKE hiredis
MAKE linenoise
MAKE hdr_histogram
MAKE lua
MAKE fpconv
MAKE jemalloc
Processing json files...
checking for xsltproc... false
checking for gcc... gcc
Linking container command to subcommands...
Checking all commands...
Generating commands.def...
All done, exiting.
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether gcc accepts -g... yes
checking for gcc option to enable C11 features... none needed
checking whether compiler is cray... no
checking whether compiler supports -std=gnu11... yes
checking whether compiler supports -Werror=unknown-warning-option... no
checking whether compiler supports -Wall... yes
checking whether compiler supports -Wextra... yes
checking whether compiler supports -Wshorten-64-to-32... no
checking whether compiler supports -Wsign-compare... yes
checking whether compiler supports -Wundef... yes
checking whether compiler supports -Wno-format-zero-length... yes
checking whether compiler supports -Wpointer-arith... yes
checking whether compiler supports -Wno-missing-braces... yes
checking whether compiler supports -Wno-missing-field-initializers... yes
checking whether compiler supports -Wno-missing-attributes... yes
checking whether compiler supports -pipe... yes
checking whether compiler supports -g3... yes
checking how to run the C preprocessor... gcc -E
checking for stdio.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for strings.h... yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for unistd.h... yes
checking whether byte ordering is bigendian... no
checking size of void *... 8
checking size of int... 4
checking size of long... 8
checking size of long long... 8
checking size of intmax_t... 8
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking whether pause instruction is compilable... yes
checking number of significant virtual address bits... 48
checking for ar... ar
checking for nm... nm
checking for gawk... no
checking for mawk... mawk
checking for malloc.h... yes
checking whether malloc_usable_size definition can use const argument... no
checking for library containing log... -lm
checking whether __attribute__ syntax is compilable... yes
checking whether compiler supports -fvisibility=hidden... yes
checking whether compiler supports -fvisibility=hidden... no
checking whether compiler supports -Werror... yes
checking whether compiler supports -herror_on_warning... yes
checking whether tls_model attribute is compilable... yes
checking whether compiler supports -Werror... yes
checking whether compiler supports -herror_on_warning... yes
checking whether alloc_size attribute is compilable... yes
checking whether compiler supports -Werror... yes
checking whether compiler supports -herror_on_warning... yes
checking whether format(gnu_printf, ...) attribute is compilable... yes
checking whether compiler supports -Werror... yes
checking whether compiler supports -herror_on_warning... yes
checking whether format(printf, ...) attribute is compilable... yes
checking whether compiler supports -Werror... yes
checking whether compiler supports -herror_on_warning... yes
checking whether format(printf, ...) attribute is compilable... yes
checking whether compiler supports -Wimplicit-fallthrough... yes
checking whether fallthrough attribute is compilable... yes
checking whether compiler supports -Wimplicit-fallthrough... yes
checking whether compiler supports -Wimplicit-fallthrough... no
checking whether compiler supports -Werror... yes
checking whether compiler supports -herror_on_warning... yes
checking whether cold attribute is compilable... yes
checking whether vm_make_tag is compilable... no
checking for a BSD-compatible install... /usr/bin/install -c
checking for ranlib... ranlib
checking for ld... /usr/bin/ld
checking for autoconf... false
checking for memalign... yes
checking for valloc... yes
checking for malloc_size... no
checking whether compiler supports -O3... yes
checking whether compiler supports -O3... no
checking whether compiler supports -funroll-loops... yes
checking configured backtracing method... N/A
checking for sbrk... yes
checking whether utrace(2) is compilable... no
checking whether utrace(2) with label is compilable... no
checking whether a program using __builtin_unreachable is compilable... yes
checking whether a program using __builtin_ffsl is compilable... yes
checking whether a program using __builtin_popcountl is compilable... yes
checking LG_PAGE... 12
checking for pthread.h... yes
checking for pthread_create in -lpthread... yes
checking for dlfcn.h... yes
checking for dlsym... yes
checking whether pthread_atfork(3) is compilable... yes
checking whether pthread_setname_np(3) is compilable... yes
checking whether pthread_getname_np(3) is compilable... yes
checking whether pthread_get_name_np(3) is compilable... no
checking for library containing clock_gettime... none required
checking whether clock_gettime(CLOCK_MONOTONIC_COARSE, ...) is compilable... yes
checking whether clock_gettime(CLOCK_MONOTONIC, ...) is compilable... yes
checking whether mach_absolute_time() is compilable... no
checking whether clock_gettime(CLOCK_REALTIME, ...) is compilable... yes
checking whether compiler supports -Werror... yes
checking whether syscall(2) is compilable... yes
checking for secure_getenv... yes
checking for sched_getcpu... yes
checking for sched_setaffinity... yes
checking for issetugid... no
checking for _malloc_thread_cleanup... no
checking for _pthread_mutex_init_calloc_cb... no
checking for memcntl... no
checking for TLS... yes
checking whether C11 atomics is compilable... yes
checking whether GCC __atomic atomics is compilable... yes
checking whether GCC 8-bit __atomic atomics is compilable... yes
checking whether GCC __sync atomics is compilable... yes
checking whether GCC 8-bit __sync atomics is compilable... yes
checking whether Darwin OSAtomic*() is compilable... no
checking whether madvise(2) is compilable... yes
checking whether madvise(..., MADV_FREE) is compilable... yes
checking whether madvise(..., MADV_DONTNEED) is compilable... yes
checking whether madvise(..., MADV_DO[NT]DUMP) is compilable... yes
checking whether madvise(..., MADV_[NO]HUGEPAGE) is compilable... yes
checking whether madvise(..., MADV_[NO]CORE) is compilable... no
checking whether mprotect(2) is compilable... yes
checking for __builtin_clz... yes
checking whether Darwin os_unfair_lock_*() is compilable... no
checking whether glibc malloc hook is compilable... no
checking whether glibc memalign hook is compilable... no
checking whether pthreads adaptive mutexes is compilable... yes
checking whether compiler supports -D_GNU_SOURCE... yes
checking whether compiler supports -Werror... yes
checking whether compiler supports -herror_on_warning... yes
checking whether strerror_r returns char with gnu source is compilable... yes
checking for _Bool... yes
checking for stdbool.h that conforms to C99... yes
configure: creating ./config.status
config.status: creating Makefile
config.status: creating jemalloc.pc
config.status: creating doc/html.xsl
config.status: creating doc/manpages.xsl
config.status: creating doc/jemalloc.xml
config.status: creating include/jemalloc/jemalloc_macros.h
config.status: creating include/jemalloc/jemalloc_protos.h
config.status: creating include/jemalloc/jemalloc_typedefs.h
config.status: creating include/jemalloc/internal/jemalloc_preamble.h
config.status: creating test/test.sh
config.status: creating test/include/test/jemalloc_test.h
config.status: creating config.stamp
config.status: creating bin/jemalloc-config
config.status: creating bin/jemalloc.sh
config.status: creating bin/jeprof
config.status: creating include/jemalloc/jemalloc_defs.h
config.status: creating include/jemalloc/internal/jemalloc_internal_defs.h
config.status: creating test/include/test/jemalloc_test_defs.h
config.status: executing include/jemalloc/internal/public_symbols.txt commands
config.status: executing include/jemalloc/internal/private_symbols.awk commands
config.status: executing include/jemalloc/internal/private_symbols_jet.awk commands
config.status: executing include/jemalloc/internal/public_namespace.h commands
config.status: executing include/jemalloc/internal/public_unnamespace.h commands
config.status: executing include/jemalloc/jemalloc_protos_jet.h commands
config.status: executing include/jemalloc/jemalloc_rename.h commands
config.status: executing include/jemalloc/jemalloc_mangle.h commands
config.status: executing include/jemalloc/jemalloc_mangle_jet.h commands
config.status: executing include/jemalloc/jemalloc.h commands
===============================================================================
jemalloc version   : 5.3.0-0-g0
library revision   : 2

CONFIG             : --disable-cxx --with-version=5.3.0-0-g0 --with-lg-quantum=3 --disable-cache-oblivious --with-jemalloc-prefix=je_ CFLAGS= LDFLAGS=
CC                 : gcc
CONFIGURE_CFLAGS   : -std=gnu11 -Wall -Wextra -Wsign-compare -Wundef -Wno-format-zero-length -Wpointer-arith -Wno-missing-braces -Wno-missing-field-initializers -Wno-missing-attributes -pipe -g3 -fvisibility=hidden -Wimplicit-fallthrough -O3 -funroll-loops
SPECIFIED_CFLAGS   : 
EXTRA_CFLAGS       : 
CPPFLAGS           : -D_GNU_SOURCE -D_REENTRANT
CXX                : 
CONFIGURE_CXXFLAGS : 
SPECIFIED_CXXFLAGS : 
EXTRA_CXXFLAGS     : 
LDFLAGS            : 
EXTRA_LDFLAGS      : 
DSO_LDFLAGS        : -shared -Wl,-soname,$(@F)
LIBS               : -lm  -pthread
RPATH_EXTRA        : 

XSLTPROC           : false
XSLROOT            : 

PREFIX             : /usr/local
BINDIR             : /usr/local/bin
DATADIR            : /usr/local/share
INCLUDEDIR         : /usr/local/include
LIBDIR             : /usr/local/lib
MANDIR             : /usr/local/share/man

srcroot            : 
abs_srcroot        : /opt/valkeyJSON/bld/_deps/valkey-src/src/valkey/deps/jemalloc/
objroot            : 
abs_objroot        : /opt/valkeyJSON/bld/_deps/valkey-src/src/valkey/deps/jemalloc/

JEMALLOC_PREFIX    : je_
JEMALLOC_PRIVATE_NAMESPACE
                   : je_
install_suffix     : 
malloc_conf        : 
documentation      : 1
shared libs        : 1
static libs        : 1
autogen            : 0
debug              : 0
stats              : 1
experimental_smallocx : 0
prof               : 0
prof-libunwind     : 0
prof-libgcc        : 0
prof-gcc           : 0
fill               : 1
utrace             : 0
xmalloc            : 0
log                : 0
lazy_lock          : 0
cache-oblivious    : 0
cxx                : 0
===============================================================================
cp include/jemalloc/internal/private_namespace.gen.h include/jemalloc/internal/private_namespace.h
    CC adlist.o
    CC ae.o
    CC anet.o
    CC dict.o
    CC kvstore.o
    CC server.o
    CC sds.o
    CC zmalloc.o
    CC threads_mngr.o
    CC lzf_c.o
    CC lzf_d.o
    CC pqsort.o
    CC zipmap.o
    CC ziplist.o
    CC sha1.o
    CC release.o
    CC io_threads.o
    CC networking.o
    CC memory_prefetch.o
    CC object.o
    CC db.o
    CC latency.o
    CC sparkline.o
    CC util.o
    CC rdb.o
    CC t_string.o
    CC t_set.o
    CC valkey-check-rdb.o
    CC valkey-check-aof.o
    CC t_zset.o
    CC t_list.o
    CC expire.o
    CC geohash.o
    CC hyperloglog.o
    CC lazyfree.o
    CC aof.o
    CC module.o
    CC pubsub.o
    CC evict.o
    CC config.o
    CC geohash_helper.o
    CC t_hash.o
    CC childinfo.o
    CC debug.o
    CC siphash.o
    CC t_stream.o
    CC listpack.o
    CC localtime.o
    CC intset.o
    CC lolwut.o
    CC lolwut5.o
    CC lolwut6.o
    CC acl.o
    CC defrag.o
    CC syncio.o
    CC tracking.o
    CC socket.o
    CC tls.o
    CC cluster.o
    CC sha256.o
    CC cluster_slot_stats.o
    CC crc16.o
    CC slowlog.o
    CC eval.o
    CC cluster_legacy.o
    CC geo.o
    CC endianconv.o
    CC setcpuaffinity.o
    CC timeout.o
    CC monotonic.o
    CC rand.o
    CC syscheck.o
    CC bio.o
    CC mt19937-64.o
    CC crcspeed.o
    CC resp_parser.o
    CC crc64.o
    CC call_reply.o
    CC memtest.o
    CC rio.o
    CC bitops.o
    CC notify.o
    CC crccombine.o
    CC setproctitle.o
    CC sentinel.o
    CC blocked.o
    CC script.o
    CC functions.o
    CC rax.o
    CC function_lua.o
    CC connection.o
    CC unix.o
    CC commands.o
    CC logreqres.o
    CC strl.o
    CC valkey-cli.o
    CC serverassert.o
    CC cli_common.o
    CC cli_commands.o
    CC valkey-benchmark.o
    CC script_lua.o
    CC replication.o
    CC sort.o
    CC multi.o
    CC quicklist.o
    LINK valkey-benchmark
    LINK valkey-cli
    LINK valkey-server
    INSTALL valkey-sentinel
    INSTALL valkey-check-rdb
    INSTALL valkey-check-aof

Hint: It's a good idea to run 'make test' ;)

[ 17%] No install step for 'valkey'
[ 19%] Completed 'valkey'
Copied valkeymodule.h and valkey-server to destination directories
[ 19%] Built target valkey
[ 21%] Creating directories for 'valkey-bloom'
[ 23%] Performing download step (git clone) for 'valkey-bloom'
Cloning into 'valkey-bloom'...
Already on 'unstable'
Your branch is up to date with 'origin/unstable'.
[ 26%] Performing update step for 'valkey-bloom'
[ 28%] No patch step for 'valkey-bloom'
[ 30%] No configure step for 'valkey-bloom'
[ 32%] No build step for 'valkey-bloom'
[ 34%] Copying pytest files to tst/integration directory
[ 36%] No install step for 'valkey-bloom'
[ 39%] Completed 'valkey-bloom'
[ 39%] Built target valkey-bloom
[ 41%] Building CXX object src/CMakeFiles/json-objects.dir/json/json.cc.o
[ 43%] Building CXX object src/CMakeFiles/json-objects.dir/json/dom.cc.o
/opt/valkeyJSON/src/json/dom.cc: In function 'JValue readLegacyDoubleAsJValue(ValkeyModuleIO*)':
/opt/valkeyJSON/src/json/dom.cc:1419:9: error: the address of 'str' will never be NULL [-Werror=address]
 1419 |     if (str) {
      |         ^~~
/opt/valkeyJSON/src/json/dom.cc:1417:10: note: 'str' declared here
 1417 |     char str[BUF_SIZE_DOUBLE_JSON];
      |          ^~~
cc1plus: all warnings being treated as errors
make[2]: *** [src/CMakeFiles/json-objects.dir/build.make:90: src/CMakeFiles/json-objects.dir/json/dom.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:315: src/CMakeFiles/json-objects.dir/all] Error 2
make: *** [Makefile:136: all] Error 2

Had the same issue 

lsb_release -a 

[build.log](https://github.com/user-attachments/files/18776143/build.log)

Distributor ID:	Debian
Description:	Debian GNU/Linux 12 (bookworm)
Release:	12
Codename:	bookworm

`/opt/valkeyJSON/src/json/dom.cc: In function 'JValue readLegacyDoubleAsJValue(ValkeyModuleIO*)':
/opt/valkeyJSON/src/json/dom.cc:1419:9: error: the address of 'str' will never be NULL [-Werror=address]
 1419 |     if (str) {
      |         ^~~
/opt/valkeyJSON/src/json/dom.cc:1417:10: note: 'str' declared here
 1417 |     char str[BUF_SIZE_DOUBLE_JSON];
      |          ^~~
cc1plus: all warnings being treated as errors
make[2]: *** [src/CMakeFiles/json-objects.dir/build.make:90: src/CMakeFiles/json-objects.dir/json/dom.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:315: src/CMakeFiles/json-objects.dir/all] Error 2
make: *** [Makefile:136: all] Error 2`

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -4,7 +4,7 @@ on: [push, pull_request]
 
 jobs:
   build-ubuntu-latest:
-    runs-on: ubuntu-22.04
+    runs-on: ubuntu-latest
     strategy:
       fail-fast: false
       matrix:

--- a/src/json/dom.cc
+++ b/src/json/dom.cc
@@ -1416,13 +1416,8 @@ STATIC JValue readLegacyDoubleAsJValue(ValkeyModuleIO *rdb) {
     double d = ValkeyModule_LoadDouble(rdb);
     char str[BUF_SIZE_DOUBLE_JSON];
     size_t str_len = jsonutil_double_to_string(d, str, sizeof(str));
-    if (str) {
         JValue v(str, str_len, allocator, false, /* isdouble */ true);
         return v;
-    } else {
-        ValkeyModule_LogIOError(rdb, "error", "Unable to read legacy double");
-        return JValue();
-    }
 }
 
 /*

--- a/src/json/memory.h
+++ b/src/json/memory.h
@@ -98,7 +98,6 @@ namespace jsn
 template <typename T> class stl_allocator {
     public:
         using value_type = T;
-
         stl_allocator() = default;
         stl_allocator(const stl_allocator<T>&) noexcept = default;
         stl_allocator(stl_allocator<T>&&) noexcept = default;

--- a/src/json/memory.h
+++ b/src/json/memory.h
@@ -95,17 +95,40 @@ namespace jsn
 //
 // Our custom allocator
 //
-template <typename T> class stl_allocator : public std::allocator<T> {
- public:
-    typedef T value_type;
-    stl_allocator() = default;
-    stl_allocator(std::allocator<T>&) {}
-    stl_allocator(std::allocator<T>&&) {}
-    template <class U> constexpr stl_allocator(const stl_allocator<U>&) noexcept {}
+template <typename T> class stl_allocator {
+    public:
+        using value_type = T;
 
-    T *allocate(std::size_t n) { return static_cast<T *>(memory_alloc(n*sizeof(T))); }
-    void deallocate(T *p, std::size_t n) { (void)n; memory_free(p); }
+        stl_allocator() = default;
+        stl_allocator(const stl_allocator<T>&) noexcept = default;
+        stl_allocator(stl_allocator<T>&&) noexcept = default;
+        template <class U>
+        constexpr stl_allocator(const stl_allocator<U>&) noexcept {}
+
+        T* allocate(std::size_t n) {
+            return static_cast<T*>(memory_alloc(n * sizeof(T)));
+        }
+
+        void deallocate(T* p, std::size_t n) {
+            (void)n; 
+            memory_free(p);
+        }
+    
+    private:
+        static void* memory_alloc(std::size_t size) {
+            return std::malloc(size);
+        }
+
+        static void memory_free(void* p) {
+            std::free(p);
+        }
 };
+
+template <typename T, typename U>
+bool operator==(const stl_allocator<T>&, const stl_allocator<U>&) { return true; }
+
+template <typename T, typename U>
+bool operator!=(const stl_allocator<T>&, const stl_allocator<U>&) { return false; }
 
 template<class Elm> using vector = std::vector<Elm, stl_allocator<Elm>>;
 

--- a/src/rapidjson/document.h
+++ b/src/rapidjson/document.h
@@ -2512,7 +2512,7 @@ private:
         DoReserveMembersVec(count, allocator);
         const typename GenericValue<Encoding,SourceAllocator>::Member* rm = rhs.GetMembersPointerVec();
         for (SizeType i = 0; i < count; i++) {
-            KeyTable_Handle name = std::move(keyTable->clone(rm[i].name));
+            KeyTable_Handle name = keyTable->clone(rm[i].name);
             GenericValue value(rm[i].value, allocator, copyConstStrings);
             DoAddMember(name, value, allocator);
         }

